### DISCUSSION TITLE
Unable to operate DIO pins on rPi 4B using Java

### DISCUSSION THREAD

**Discussion by plsherman** [Pi4J/pi4j#380] [Q&A] [answered]
Can anyone provide me with suggestions to resolve these issues?

I'm trying to run the sample digital output program on a new Pi4 and have had no success. The program appears to run using sudo jbang, reporting all of the pin state changes but the pin never changes state. Jbang also works the same if I run it without using sudo which indicates that it's not using the libpigpio.so module which requires sudo.   If I run it standalone, it fails with two errors. I can fix the first using a path in the -D system property option but the second remains. Standalone invocation fails the same way when also invoked using sudo.

The OS on the Pi is the downloaded Pi4J-Basic-OS-main.img.zip

When I invoked the program using jbang, I added code to verify what I/O providers were being used. The output showed that PiGpio was the I/O provider. The jbang invocation:
sudo  `which jbang`  DigitalOutputFromProperties

jbang additions to the program copied from a jbang example:
///usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS org.slf4j:slf4j-api:1.7.35
//DEPS org.slf4j:slf4j-simple:1.7.35
//DEPS com.pi4j:pi4j-core:2.3.0
//DEPS com.pi4j:pi4j-plugin-raspberrypi:2.3.0
//DEPS com.pi4j:pi4j-plugin-pigpio:2.3.0

Standalone invocation was done supplying the class file name using my pi4jJava single line script:
java -classpath .:/usr/lib/pi4j2.6.1:/usr/lib/pi4j2.6.1/'*' $*

The first error is a failure to load the libgpiod.so. I can eliminate this using the java.library.path system property. When I ran ldconfig to add this library to the cache, it changed the name of the executable libgpiod.so to libgpiod.so.2 so I added the standard alias without the version number. This same error occurred before I ran ldconfig where the executable was contained in the libgpiod.so file. I've included ls outputs and the contents of the library cache.

[main] ERROR com.pi4j.library.gpiod.util.NativeLibraryLoader - Unable to load [libgpiod.so] using path: [/lib/aarch64/pi4j-gpiod/libgpiod.so]
java.lang.UnsatisfiedLinkError: /tmp/libgpiod8301055695622490843.so: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /tmp/libgpiod8301055695622490843.so)

ls -l /lib/aarch64/pi4j-gpiod/libgpiod.s*
lrwxrwxrwx 1 root root 13 Sep  5 22:39 /lib/aarch64/pi4j-gpiod/libgpiod.so -> libgpiod.so.2
-rwxr-xr-x 1 root root 126672 Sep  5 12:17 /lib/aarch64/pi4j-gpiod/libgpiod.so.2

ldconfig -p | grep libgpiod
libgpiod.so.2 (libc6,AArch64) => /lib/aarch64/pi4j-gpiod/libgpiod.so.2
libgpiod.so (libc6,AArch64) => /lib/aarch64/pi4j-gpiod/libgpiod.so

Program execution continues then stops after the following error:

Exception in thread "main" java.lang.UnsatisfiedLinkError: Pi4J was unable to extract and load the native library [/lib/aarch64/pi4j-gpiod/libgpiod.so] from the embedded resources inside this JAR [/usr/lib/pi4j2.6.1/pi4j-library-gpiod.jar]. to a temporary location on this system.  You can alternatively define the 'pi4j.library.path' system property to override this behavior and specify the library path.

When I run the program using the command:
      sudo  pi4jJava -Dpi4j.library.path="/lib/aarch64/pi4j-gpiod" PwmMotor
I still get the extract failure. I also ran the program with all parameters supplied on the command line, bypassing my script, and had the same issue.

**Comment by taartspi** [comment]
You show several different test programs you tried.  Referring to only the jbang minimal example.    
Using a pi4j with bookworm.      Followed instructions on  https://www.pi4j.com/examples/jbang/jbang_minimal_example/      
Used link to install jbang  0.117.1   
Created a new dir and within Pi4JMinimalExample.java     
Copied contents from that example, making no changes   
Connected LED to GPIO22  (pin 15)  
Enter:  sudo `which jbang` Pi4JMinimalExample.java      The LED blinks.        

Can you confirm what results you see doing these same steps.   
Confirm your OS is  Latest Pi4J-Basic-OS Image: [Pi4J-Basic-OS-main.img.zip (1.88 GiB, 2023-07-31 11:45)](https://pi4j-download.com/Pi4J-Basic-OS-main.img.zip)      

With those answers I can see if I need to setup a SD card with that Basic Image and test those minimal steps again.

**Reply by plsherman** [reply]
It appears that I don't have the correct version of the OS image. This is what I downloaded from the web site and what it contained. The img file doesn't have the same date and time that you said it should have. The .img file was extracted from the .zip file
ls -l Pi4J-Basic-OS-main.img.zip *.img
-rw-r--r--. 1 root     root  6442450944 Jul 29  2114 Pi4J-Basic-OS.img
-rw-r--r--. 1 psherman users 2022384795 Aug 29 13:00 Pi4J-Basic-OS-main.img.zip

I downloaded the minimal example and attempted to run it using jbang.
jbang --version
0.117.1

jbang Pi4JMinimalExample.java
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
[main] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
Exception in thread "main" java.lang.reflect.UndeclaredThrowableException
	at jdk.proxy1/jdk.proxy1.$Proxy2.create(Unknown Source)
	at com.pi4j.context.Context.create(Context.java:317)
	at com.pi4j.internal.IOCreator.create(IOCreator.java:58)
	at com.pi4j.internal.IOCreator.create(IOCreator.java:96)
	at com.pi4j.internal.IOCreator.create(IOCreator.java:176)
	at Pi4JMinimalExample.main(Pi4JMinimalExample.java:38)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at com.pi4j.provider.impl.ProviderProxyHandler.invoke(ProviderProxyHandler.java:100)
	... 6 more
Caused by: com.pi4j.library.pigpio.PiGpioException: PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
	at com.pi4j.library.pigpio.impl.PiGpioBase.validateResult(PiGpioBase.java:265)
	at com.pi4j.library.pigpio.impl.PiGpioBase.validateResult(PiGpioBase.java:251)
	at com.pi4j.library.pigpio.impl.PiGpioNativeImpl.gpioInitialise(PiGpioNativeImpl.java:98)
	at com.pi4j.library.pigpio.PiGpio.initialize(PiGpio.java:159)
	at com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl.create(PiGpioDigitalOutputProviderImpl.java:60)
	at com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl.create(PiGpioDigitalOutputProviderImpl.java:41)
	... 11 more


I successfully compiled the program using my javac procedure the attempted to run it using an equivalent java procedure.
pi4jJava  Pi4JMinimalExample
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.boardinfo.util.BoardInfoHelper - Detected OS: Name: Linux, version: 6.1.21-v8+, architecture: aarch64
[main] INFO com.pi4j.boardinfo.util.BoardInfoHelper - Detected Java: Version: 17.0.12, runtime: 17.0.12+7-Debian-2deb11u1, vendor: Debian, vendor version: null
[main] INFO com.pi4j.boardinfo.util.BoardInfoHelper - Detected board type MODEL_4_B by code: d03115
[main] INFO com.pi4j.context.impl.DefaultContext - Detected board model: Raspberry Pi 4 Model B
[main] INFO com.pi4j.context.impl.DefaultContext - Running on: Name: Linux, version: 6.1.21-v8+, architecture: aarch64
[main] INFO com.pi4j.context.impl.DefaultContext - With Java version: Version: 17.0.12, runtime: 17.0.12+7-Debian-2deb11u1, vendor: Debian, vendor version: null
[main] INFO com.pi4j.runtime.impl.DefaultRuntime - Initializing Pi4J context/runtime...
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider DIGITAL_OUTPUT RaspberryPi Digital Output (GPIO) Provider with priority 0 with provider GpioD Digital Output (GPIO) Provider with higher priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider DIGITAL_INPUT RaspberryPi Digital Input (GPIO) Provider with priority 0 with provider GpioD Digital Input (GPIO) Provider with higher priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Ignoring provider DIGITAL_OUTPUT LinuxFS Digital Output (GPIO) Provider with priority 50 as lower priority than GpioD Digital Output (GPIO) Provider which has priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider I2C RaspberryPi I2C Provider with priority 0 with provider LinuxFS I2C Provider with higher priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider PWM RaspberryPi PWM Provider with priority 0 with provider LinuxFS PWM Provider with higher priority 50
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Ignoring provider DIGITAL_INPUT LinuxFS Digital Input (GPIO) Provider with priority 50 as lower priority than GpioD Digital Input (GPIO) Provider which has priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider SERIAL RaspberryPi Serial Provider with priority 0 with provider PiGpio Serial Provider with higher priority 100
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Ignoring provider I2C PiGpio I2C Provider with priority 100 as lower priority than LinuxFS I2C Provider which has priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider SPI RaspberryPi SPI Provider with priority 0 with provider PiGpio SPI Provider with higher priority 100
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Ignoring provider DIGITAL_INPUT PiGpio Digital Input (GPIO) Provider with priority 100 as lower priority than GpioD Digital Input (GPIO) Provider which has priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Ignoring provider DIGITAL_OUTPUT PiGpio Digital Output (GPIO) Provider with priority 100 as lower priority than GpioD Digital Output (GPIO) Provider which has priority 150
[main] WARN com.pi4j.runtime.impl.DefaultRuntime - Replacing provider PWM LinuxFS PWM Provider with priority 50 with provider PiGpio PWM Provider with higher priority 100
[main] ERROR com.pi4j.library.gpiod.util.NativeLibraryLoader - Unable to load [libgpiod.so] using path: [/lib/aarch64/pi4j-gpiod/libgpiod.so]
java.lang.UnsatisfiedLinkError: /tmp/libgpiod3028389313752574261.so: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /tmp/libgpiod3028389313752574261.so)
	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
        ............
	at com.pi4j.Pi4J.newAutoContext(Pi4J.java:73)
	at Pi4JMinimalExample.main(Pi4JMinimalExample.java:29)
Exception in thread "main" java.lang.UnsatisfiedLinkError: Pi4J was unable to extract and load the native library [/lib/aarch64/pi4j-gpiod/libgpiod.so] from the embedded resources inside this JAR [/usr/lib/pi4j-2.6.1/lib/pi4j-library-gpiod.jar]. to a temporary location on this system.  You can alternatively define the 'pi4j.library.path' system property to override this behavior and specify the library path.
 UNDERLYING EXCEPTION: [java.lang.UnsatisfiedLinkError]=/tmp/libgpiod3028389313752574261.so: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /tmp/libgpiod3028389313752574261.so)
	at com.pi4j.library.gpiod.util.NativeLibraryLoader.load(NativeLibraryLoader.java:201)
        ..........
	at com.pi4j.Pi4J.newAutoContext(Pi4J.java:73)
	at Pi4JMinimalExample.main(Pi4JMinimalExample.java:29)
[pi4j-shutdown] INFO com.pi4j.runtime.impl.DefaultRuntime - Shutting down Pi4J context/runtime...
[pi4j-shutdown] INFO com.pi4j.util.ExecutorPool - Shutting down executor pool Pi4J.RUNTIME
[pi4j-shutdown] INFO com.pi4j.runtime.impl.DefaultRuntime - Pi4J context/runtime successfully shutdown. Dispatching shutdown event.

**Comment by taartspi** [comment]
The zip I listed was found from the web pages.  I do not use these for my OS so I can't comment on different dates. But as this program is simple I would expect any of the recent official zips would work.   
The Pigpio Init error.  I do not see sudo in front of the jbang command  use   sudo which jbang Pi4JMinimalExample.java   

When you tried using the example as a regular java program, you are then using your POM file dependencies which appear to include gpiod. You should remove gpiod dependency so the PiGpio provider is not replaced;   
see:  https://www.pi4j.com/documentation/providers/

**Reply by plsherman** [reply, marked as answer]
Thank you for your assistance. Between it and additional research I've been able to get both Python And Java to use the hardware PWM pins. Python uses the hardware as specified but it appears that Java does not.  When I set up (java) the hardware PWM pin using a config variable with frequency set to 250, I ended up with a jittery frequency of 167Hz, what I'd expect from a software PWM. My Python tests showed that the HW frequency should be as specified and show no jitter.

*********  LATE NIGHT UPDATE  -  ISSUE RESOLVED  ****************
I was correct - My program was not using the hardware PWM! My suspicion is that it's a Pi4J issue involving mishandling the data provided in the config variable. I specified hardware and a default frequency. This set up a software PWM instead of the hardware version. I added a "pwm.frequency(250)" statement before the pwm.on() call and the pin switched to the hardware driven PWM. Without doing additional experimentation, my guess is that the config frequency variable was never processed causing the hardware PWM to not be used. 

I'm an advisor (and lead) to a high school FIRST Robotics team that's building a motorized mobile recliner as an off season project. We have two drive trains whose motors are controlled using PWM, a situation that's a perfect fit for rPi control. I can now have my cadre of newbies start learning java by designing and writing the code to control the chair.

Thanks again for your assistance
Phil

**Reply by FDelporte** [reply]
@plsherman your FIRST project sounds like a perfect story for https://www.pi4j.com/featured-projects/. If possible, please reach out to me via frank@pi4j.com with some more info and pictures.

**Comment by eitch** [comment]
@plsherman do you have some example code, so we can look at the issue? Is it a documentation error or a bug? =)
